Marlin/src/feature/powerloss.cpp

```diff
@@ -430,9 +435,7 @@ void PrintJobRecovery::resume() {
   gcode.process_subcommands_now(cmd);
 
   // Un-retract
-  #if POWER_LOSS_PURGE_LEN
-    //sprintf_P(cmd, PSTR("G1 E%d F3000"), POWER_LOSS_PURGE_LEN);
-    //gcode.process_subcommands_now(cmd);
+  #if POWER_LOSS_RETRACT_LEN
```

Comment:
This change is slightly confusing. It looks like a mistake at first glance, since I check for `POWER_LOSS_RETRACT_LEN`, then move to `POWER_LOSS_PURGE_LEN`.

This is because this move is undoing a retract, and is not necessary if the retract was 0. The purge length is simply the destination of the move.

This file guarantees that both these constants are defined, so they can be readily used without actually checking them for presence. The `#if` checks are just there to bypass no-op moves.

Reply:
Looking over the resume behavior, I'm not sure the purge and un-retract are being applied completely.

First, note that the E position is reset to 0 at the start of `resume`, and not to `-(POWER_LOSS_RETRACT_LEN)`. Following common norms, on resume there would be an un-retract of `+(POWER_LOSS_RETRACT_LEN)` followed later by an additional purge of `POWER_LOSS_PURGE_LEN`.

The way that the settings are described it's ambiguous whether the `POWER_LOSS_RETRACT_LEN` is undone on `resume`, or whether `POWER_LOSS_PURGE_LEN` is supposed to include it.

The bit above where it says "Extrude and retract to clean the nozzle" is fanciful. We don't have any settings for "cleaning the nozzle" such as proper ramming sequence. All we have is a length of retract on power outage (usually zero) and some length of extra purge on resume (to replace oozed material). If we want a proper nozzle cleaning with `G12` this would be the place to add that, but for now these two settings should be applied as literally as possible.

So, I propose that the un-retract be placed first. And then the additional purge. And then nothing else.

Reply:
I see you are already doing this, so I'll step back unless you ask me to make any more changes to it.